File: Gate.py
class Gate:
    def __init__(self, nr_inputs, nr_outputs) -> None:
        self.nr_inputs = nr_inputs
        self.nr_outputs = nr_outputs
        self.inputs = {}
        self.outputs = {}
    
    def appendInput(self, b):
        self.inputs[len(list(self.inputs.keys()))] = b
    
    def appendOutput(self, b):
        self.outputs[len(list(self.outputs.keys()))] = b

    def recieve(self, value, x):
        # rerun the processing
        print("recieved ", value, x)
        self.send(value)
    
    def send(self, y):
        for x in list(self.outputs.keys()):
            self.outputs[x].recieveLeft(y)

File: javascriptPort/src/library/gate.ts
import uuid from "uuid"
import Bus from "./bus";

class Gate {
    nr_inputs: number;
    nr_outputs: number;
    type: null | string;
    id: string;
    inputs: {
        [key: string]: Bus
    };
    inp_values: Array<number | null | string | Gate>;
    outputs: {
        [key: string]: Bus
    } 

    constructor(nr_inputs: number, nr_outputs: number, type = null) {
        this.nr_inputs = nr_inputs;
        this.nr_outputs = nr_outputs;
        this.type = type;
        this.inputs = {};
        this.inp_values = [];
        this.outputs = {};
        this.id = uuid.v4();
    }

    appendInput(b: Bus) {
        b.recieveLeft(this);
        this.inputs[Object.keys(this.inputs).length] = b;
    }

    appendOutput(b: Bus) {
        this.outputs[Object.keys(this.outputs).length] = b;
    }

    recieve(value: (number | Gate)) {
        this.inp_values.push(value);
        if (this.inp_values.length < this.nr_inputs) {
            this.send(value);
        }
    }

    send(y: number | Gate) {
        Object.keys(this.outputs).forEach(x => {
            this.outputs[x].recieveLeft(y);
        });
    }

    toJson() {
        return {
            "leftNodes": Object.keys(this.inputs).map(x => this.inputs[x]).map(x => x.id),
            "rightNodes": Object.keys(this.outputs).map(x => this.outputs[x]).map(x => x.id),
            "type": this.type,
            "id": this.id
        };
    }
}

export default Gate;